src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -1,72 +1,112 @@
+                var diagnosticAnalyzer = _owner._analyzerService.CreateIncrementalAnalyzer(_workspace);
 
-                var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
-                var source = s_cancellationTokens.GetValue(workspace, _ => new CancellationTokenSource());
+                var solution = _workspace.CurrentSolution;
+                var documentIds = _workspace.GetOpenDocumentIds().ToImmutableArray();
 
-                var solution = workspace.CurrentSolution;
-                foreach (var documentId in workspace.GetOpenDocumentIds())
+                try
                 {
-                    var document = solution.GetDocument(documentId);
-                    if (document == null)
+                    foreach (var documentId in documentIds)
                     {
-                        continue;
-                    }
+                        var document = solution.GetDocument(documentId);
+                        if (document == null)
+                        {
+                            continue;
+                        }
 
-                    // delay analyzing
-                    await Task.Delay(workerBackOffTimeSpanInMS).ConfigureAwait(false);
+                        // delay analyzing
+                        await Task.Delay(workerBackOffTimeSpanInMS, _source.Token).ConfigureAwait(false);
 
-                    // do actual analysis
-                    await analyzer.AnalyzeSyntaxAsync(document, source.Token).ConfigureAwait(false);
-                    await analyzer.AnalyzeDocumentAsync(document, bodyOpt: null, cancellationToken: source.Token).ConfigureAwait(false);
+                        // do actual analysis
+                        await diagnosticAnalyzer.AnalyzeSyntaxAsync(document, _source.Token).ConfigureAwait(false);
+                        await diagnosticAnalyzer.AnalyzeDocumentAsync(document, bodyOpt: null, cancellationToken: _source.Token).ConfigureAwait(false);
 
-                    // don't call project one.
+                        // don't call project one.
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // do nothing
                 }
             }
-            catch (OperationCanceledException)
-            {
-                // do nothing
-            }
-        }
 
-        public void Unregister(Workspace workspace, bool blockingShutdown = false)
-        {
-            CancellationTokenSource source;
-            if (s_cancellationTokens.TryGetValue(workspace, out source))
+            public async void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
-                source.Cancel();
+                Contract.ThrowIfFalse(workspace == _workspace);
+                _source.Cancel();
+
+                // wait for analyzer work to be finished
+                await _analyzeTask.ConfigureAwait(false);
```

Comment:
thought about that, for preview workspace, I think it is fine since as long as it gets cleaned up. that is all matter. for caller's point of view, it shouldn't care.